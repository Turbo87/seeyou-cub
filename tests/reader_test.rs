use seeyou_cub::{CubReader, CubStyle};

#[test]
fn parse_france_fixture() {
    let mut reader = CubReader::from_path("tests/fixtures/france_2024.07.02.cub")
        .expect("Failed to open fixture file");
    let mut warnings = Vec::new();

    let header = reader
        .read_header(&mut warnings)
        .expect("Failed to parse header");

    // Validate warnings
    assert_eq!(warnings.len(), 0);

    // Snapshot the header
    insta::assert_debug_snapshot!("header", header);

    // Parse all items
    let items: Vec<_> = reader
        .read_items(&header, &mut warnings)
        .expect("Failed to create item iterator")
        .collect::<Result<Vec<_>, _>>()
        .expect("Failed to parse items");

    // Validate item count
    assert_eq!(items.len(), 1368);

    // Snapshot first few items
    insta::assert_debug_snapshot!("items_sample", &items[0..5]);

    // Validate first item's points
    let first_item = items.first().unwrap();
    let points: Vec<_> = reader
        .read_points(&header, first_item, &mut warnings)
        .expect("Failed to read points")
        .collect::<Result<Vec<_>, _>>()
        .expect("Failed to parse points");

    insta::assert_debug_snapshot!("first_item_points", points);

    let last_item = items.last().unwrap();
    let points: Vec<_> = reader
        .read_points(&header, last_item, &mut warnings)
        .expect("Failed to read points")
        .collect::<Result<Vec<_>, _>>()
        .expect("Failed to parse points");

    insta::assert_debug_snapshot!("last_item_points", points);

    // Validate all airspaces and collect statistics
    let mut total_points = 0;
    let mut style_counts = std::collections::HashMap::new();

    for item in &items {
        *style_counts.entry(item.style()).or_insert(0) += 1;

        let mut point_iter = reader
            .read_points(&header, item, &mut warnings)
            .expect("Failed to read points");

        for point_result in &mut point_iter {
            let _point = point_result.expect("Failed to parse point");
            total_points += 1;
        }
    }

    // Validate total points
    assert_eq!(total_points, 93784, "Total points mismatch");

    // Validate airspace type distribution
    assert_eq!(
        *style_counts.get(&CubStyle::Unknown).unwrap_or(&0),
        452,
        "Unknown airspace count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::RestrictedArea).unwrap_or(&0),
        435,
        "Restricted area count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::ProhibitedArea).unwrap_or(&0),
        113,
        "Prohibited area count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::DangerArea).unwrap_or(&0),
        111,
        "Danger area count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::GliderSector).unwrap_or(&0),
        135,
        "Glider sector count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::ControlZone).unwrap_or(&0),
        92,
        "Control zone count mismatch"
    );
}
