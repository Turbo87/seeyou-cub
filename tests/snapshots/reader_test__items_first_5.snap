---
source: tests/reader_test.rs
expression: "&items[0..5]"
---
[
    Item {
        left: 0.033180647,
        top: 0.83466476,
        right: 0.033224326,
        bottom: 0.8346504,
        type_byte: 2,
        alt_style_byte: 33,
        min_alt: 0,
        max_alt: 488,
        points_offset: 0,
        time_out: 0,
        extra_data: 0,
        active_time: 0,
        extended_type_byte: 0,
    },
    Item {
        left: 0.039682,
        top: 0.85236555,
        right: 0.039759766,
        bottom: 0.85232186,
        type_byte: 3,
        alt_style_byte: 33,
        min_alt: 0,
        max_alt: 610,
        points_offset: 40,
        time_out: 0,
        extra_data: 0,
        active_time: 0,
        extended_type_byte: 0,
    },
    Item {
        left: 0.099705316,
        top: 0.7667715,
        right: 0.099785745,
        bottom: 0.76671344,
        type_byte: 3,
        alt_style_byte: 49,
        min_alt: 0,
        max_alt: 5944,
        points_offset: 80,
        time_out: 0,
        extra_data: 0,
        active_time: 0,
        extended_type_byte: 0,
    },
    Item {
        left: 0.093006656,
        top: 0.8027499,
        right: 0.09311798,
        bottom: 0.8026769,
        type_byte: 2,
        alt_style_byte: 33,
        min_alt: 0,
        max_alt: 1067,
        points_offset: 1005,
        time_out: 0,
        extra_data: 0,
        active_time: 0,
        extended_type_byte: 0,
    },
    Item {
        left: 0.03643375,
        top: 0.851842,
        right: 0.03662763,
        bottom: 0.8517546,
        type_byte: 2,
        alt_style_byte: 33,
        min_alt: 0,
        max_alt: 457,
        points_offset: 1052,
        time_out: 0,
        extra_data: 0,
        active_time: 0,
        extended_type_byte: 0,
    },
]
